executor/analyze.go

```diff
@@ -496,22 +527,465 @@ type AnalyzeFastTask struct {
+			e.collectors[len(e.colsInfo)+hasPKInfo+j].Samples[samplePos] = &statistics.SampleItem{}
+		}
+		e.collectors[len(e.colsInfo)+hasPKInfo+j].Samples[samplePos].Ordinal = int(samplePos)
+		e.collectors[len(e.colsInfo)+hasPKInfo+j].Samples[samplePos].Value = types.NewBytesDatum(bytes)
+	}
+	return nil
+}
+
+func (e *AnalyzeFastExec) handleBatchGetResponse(kvMap map[string][]byte) (err error) {
+	length := int32(len(kvMap))
+	newCursor := atomic.AddInt32(&e.sampCursor, length)
+	hasPKInfo := 0
+	if e.pkInfo != nil {
+		hasPKInfo = 1
+	}
+	samplePos := newCursor - length
+	for sKey, sValue := range kvMap {
+		err = e.updateCollectorSamples(sValue, kv.Key(sKey), samplePos, hasPKInfo)
+		if err != nil {
+			return err
+		}
+		samplePos++
+	}
+	return nil
+}
+
+func (e *AnalyzeFastExec) handleScanIter(iter kv.Iterator) (err error) {
+	hasPKInfo := 0
+	if e.pkInfo != nil {
+		hasPKInfo = 1
+	}
+	rander := rand.New(rand.NewSource(RandSeed + int64(e.rowCount)))
+	for ; iter.Valid() && err == nil; err = iter.Next() {
+		// reservoir sampling
+		e.rowCount++
+		randNum := rander.Int63n(int64(e.rowCount))
+		if randNum > int64(MaxSampleSize) && e.sampCursor == int32(MaxSampleSize) {
+			continue
+		}
+
+		p := rander.Int31n(int32(MaxSampleSize))
+		if e.sampCursor < int32(MaxSampleSize) {
+			p = e.sampCursor
+			e.sampCursor++
+		}
+
+		err = e.updateCollectorSamples(iter.Value(), iter.Key(), p, hasPKInfo)
+		if err != nil {
+			return err
+		}
+	}
+	return err
+}
+
+func (e *AnalyzeFastExec) handleScanTasks(bo *tikv.Backoffer) error {
+	snapshot, err := e.ctx.GetStore().(tikv.Storage).GetSnapshot(kv.MaxVersion)
+	if err != nil {
+		return err
+	}
+	for _, t := range e.scanTasks {
+		iter, err := snapshot.Iter(t.StartKey, t.EndKey)
+		if err != nil {
+			return err
+		}
+		err = e.handleScanIter(iter)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (e *AnalyzeFastExec) handleSampTasks(bo *tikv.Backoffer, rid int, err *error) {
+	defer func() {
+		e.wg.Done()
+	}()
+	var snapshot kv.Snapshot
+	snapshot, *err = e.ctx.GetStore().(tikv.Storage).GetSnapshot(kv.MaxVersion)
+	rander := rand.New(rand.NewSource(RandSeed + int64(rid)))
+	if *err != nil {
+		return
+	}
+	for i := rid; i < len(e.sampTasks); i += e.concurrency {
+		task := e.sampTasks[i]
+		if task.SampSize == 0 {
+			continue
+		}
+
+		var tableID, minRowID, maxRowID int64
+		startKey, endKey := task.Location.StartKey, task.Location.EndKey
+		tableID, minRowID, *err = tablecodec.DecodeRecordKey(startKey)
+		if *err != nil {
+			return
+		}
+		_, maxRowID, *err = tablecodec.DecodeRecordKey(endKey)
+		if *err != nil {
+			return
+		}
+		if maxRowID <= minRowID {
+			continue
+		}
+
+		keys := make([]kv.Key, 0, task.SampSize)
+		for i := 0; i < int(task.SampSize); i++ {
+			randKey := rander.Int63n(maxRowID-minRowID) + minRowID
+			keys = append(keys, tablecodec.EncodeRowKeyWithHandle(tableID, randKey))
+		}
+
+		var kvMap map[string][]byte
+		kvMap, *err = snapshot.BatchGet(keys)
```

Comment:
It is likely that we get no keys if the key space is quite large?

Reply:
It depends on how `BatchGet` realized. I remembered somebody said it realized by `Seek`?

Reply:
Seems that it is not currently.

Reply:
Let me have a confirmation.

Reply:
Now, it uses the `Iter`.